### PR TITLE
Change IoVecBuffer[Mut] len to u32

### PR DESCRIPTION
**Pull request by brandonpike**
## Changes
- Change IoVecBuffer[Mut] len to u32 (was usize)
- Add u32 overflow check (errors out rather than panics)

Closes https://github.com/firecracker-microvm/firecracker/issues/4548

## Reason
Protect against overflows and use a consistent data type for virtio files.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following Developer
Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this
  PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet
  [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

______________________________________________________________________

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Comment by brandonpike** [comment]
Fixing lint/build errors...

**Comment by brandonpike** [comment]
## Local Tests on R7g.metal:
### Checkstyle
```
./tools/devtool -y checkstyle
======================================================================= 13 passed in 22.86s =======================================================================
[Firecracker devtool 2024-04-24T22:10:26+00:00] Finished test run ...
```

### Integration tests
```
./tools/devtool -y test -- ../tests/integration_tests/build/
======================================================= 8 passed, 1 skipped, 1 warning in 384.33s (0:06:24) =======================================================
[Firecracker devtool 2024-04-24T22:08:11+00:00] Finished test run ...

Warning:
/firecracker/tests/integration_tests/build/test_coverage.py:106: UserWarning: Not uploading coverage report due to missing CODECOV_TOKEN environment variable
    warnings.warn(
```

### Performance Tests
```
./tools/devtool -y test --performance -c 1-10 -m 0 -- ../tests/integration_tests/ -m 'no_block_pr and not nonci' --log-cli-level=INFO
FAILED
```

Performance tests fail on R7g because of the graviton 3 processors having a different cpu register set?

**Review by roypat — COMMENTED** [review]
Hi @brandonpike, 
Thanks for picking this up! The changes generally look good, but I think we can go a bit further in the vsock module. The specification states that everything size related about vsock buffers is u32, so we can do the following changes to function signatures:

- `VsockConnection::peer_avail_credit` can return `u32`
- `VsockPacket::buf_size` can return `u32`
- `VsockPacket::read_at_offset_from` should be `(&self, &mut T, u32, u32) -> Result<u32, VsockError>` (e..g. the conversion to usize should only happen right around the `write_volatile_at` call
- ditto for `VsockPacket::write_from_offset_to`
Although, maybe we can even get away with changing the arguments `IoVecBuffer[Mut]::{read,write}_volatile_at` to `u32`s?

That being said, the changes in this PR as-is are already a huge improvement, so am also happy to approve/merge and track the above as a follow up. Just please fix the errors in the `cfg(kani)` modules :)

File: src/vmm/src/devices/virtio/net/tap.rs [hunk 1]
```diff
@@ -363,7 +363,7 @@ pub mod tests {
 
         tap.write_iovec(&scattered).unwrap();
 
-        let mut read_buf = vec![0u8; scattered.len()];
+        let mut read_buf = vec![0u8; scattered.len().try_into().unwrap()];
```
[the remark this answers is not in the document]
Comment (brandonpike) [reply src/vmm/src/devices/virtio/net/tap.rs]: Nope, the largest lossless usize conversion would be u16 -> usize. I actually prefer the primitive conversion with 'as'. 

```
let mut read_buf = vec![0u8; scattered.len() as usize];
```

[the remark this answers is not in the document]
Comment (roypat) [reply src/vmm/src/devices/virtio/net/tap.rs]: Yeah, same

**Review by roypat — COMMENTED** [review]
Thanks, looks good now! As one last thing, could you squash the fixup commit into 2c51bdd428c0a4d4cfcc9c4f87998de32ccba709? Then we should be good to merge :)

**Review by roypat — APPROVED** [review]

**Review by ShadowCurse — APPROVED** [review]

**Pull request merged**